server/msgpack/server.go

```diff
@@ -0,0 +1,240 @@
+	clockOpts := opts.ClockOptions()
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope().SubScope("server")
+	queue := packet.NewQueue(opts.PacketQueueSize(), clockOpts, instrumentOpts.SetMetricsScope(scope))
+	processor := packet.NewProcessor(queue, aggregator, opts.WorkerPoolSize(), clockOpts, instrumentOpts)
+	s := &server{
+		address:      address,
+		opts:         opts,
+		log:          instrumentOpts.Logger(),
+		iteratorPool: opts.IteratorPool(),
+		queue:        queue,
+		processor:    processor,
+		metrics:      newServerMetrics(scope),
+	}
+
+	// Set up the connection functions
+	s.addConnectionFn = s.addConnection
+	s.removeConnectionFn = s.removeConnection
+	s.handleConnectionFn = s.handleConnection
+
+	// Start reporting metrics
+	go s.reportMetrics()
+
+	return s
+}
+
+func (s *server) ListenAndServe() (xclose.SimpleCloser, error) {
+	listener, err := net.Listen("tcp", s.address)
+	if err != nil {
+		return nil, err
+	}
+	s.listener = listener
+	go s.serve()
+	return s, nil
+}
+
+func (s *server) serve() error {
+	connCh, errCh := xnet.StartAcceptLoop(s.listener, s.opts.Retrier())
+	for conn := range connCh {
+		if !s.addConnectionFn(conn) {
+			conn.Close()
+		} else {
+			s.wgConns.Add(1)
+			go s.handleConnectionFn(conn)
+		}
+	}
+	return <-errCh
+}
+
+func (s *server) Close() {
+	s.Lock()
+	if atomic.LoadInt32(&s.closed) == 1 {
+		s.Unlock()
+		return
+	}
+	atomic.StoreInt32(&s.closed, 1)
+	openConns := make([]net.Conn, len(s.conns))
+	copy(openConns, s.conns)
+	s.Unlock()
+
+	// Close all open connections
+	for _, conn := range openConns {
+		conn.Close()
+	}
+
+	// Close the listener
+	if s.listener != nil {
+		s.listener.Close()
+	}
+
+	// Wait for all connection handlers to finish
+	s.wgConns.Wait()
+
+	// Close the packet queue
+	s.queue.Close()
+
+	// Close the packet processor
+	s.processor.Close()
+}
+
+func (s *server) addConnection(conn net.Conn) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	if atomic.LoadInt32(&s.closed) == 1 {
+		return false
+	}
+	s.conns = append(s.conns, conn)
+	atomic.AddInt32(&s.numConns, 1)
+	return true
+}
+
+func (s *server) removeConnection(conn net.Conn) {
+	s.Lock()
+	defer s.Unlock()
+
+	numConns := len(s.conns)
+	for i := 0; i < numConns; i++ {
+		if s.conns[i] == conn {
+			// Move the last connection to i and reduce the number of connections by 1
+			s.conns[i] = s.conns[numConns-1]
+			s.conns = s.conns[:numConns-1]
+			atomic.AddInt32(&s.numConns, -1)
+			return
+		}
+	}
+}
+
+func (s *server) handleConnection(conn net.Conn) {
+	defer func() {
+		conn.Close()
+		s.removeConnectionFn(conn)
+		s.wgConns.Done()
+	}()
+
+	it := s.iteratorPool.Get()
+	it.Reset(bufio.NewReader(conn))
+	defer it.Close()
+
+	// Iterate over the incoming metrics stream and queue up metrics
+	for it.Next() {
+		metric, policies := it.Value()
+		if err := s.queue.Enqueue(packet.Packet{Metric: metric, Policies: policies}); err != nil {
```

Comment:
I feel the packet queue and the packet processor provide a good level of isolation between the producer (packets coming in) and the consumer (the aggregation layer).

By enqueuing the packets to the queue, this provides load shedding when the queue is overloaded. Additionally, when the queue is overwhelmed and needs to drop packets, the semantics is such that it'll drop the oldest packets and make room for the recent packets, which is important because the recent data are more valuable for monitoring and alerting.

The processor provides the ability to limit the parallelism for packet processing, so that the number of package processing goroutines is no more than the number of cores, avoiding the high context switching overhead when we have one goroutine per connection. Additionally, if one goroutine is temporarily slow when processing packets, other goroutines can pick up the slack without blocking packets coming in from any particular connection.

If the queue turns out to be a hotspot during testing, it doesn't require much change to connect the server directly to the aggregator, but for now I'd rather keep them for the reasons mentioned above unless I have more production data to show otherwise.